koma-tests/test/koma/RandomTests.kt

```diff
@@ -0,0 +1,187 @@
+        // Check that each bit is equally likely to be a 0 or 1.
+        
+        val count = 10000
+        val bitCounts = IntArray(32)
+        val rng = getRng()
+        for (i in 1..count) {
+            val value = rng.nextInt()
+            for (j in 0 until 32)
+                if (value and (1 shl j) != 0)
+                    bitCounts[j]++
+        }
+        val tol = 4*sqrt(count/2)
+        for (i in 0 until 32) {
+            assert(bitCounts[i] >= count/2 - tol)
+            assert(bitCounts[i] <= count/2 + tol)
+        }
+    }
+
+    @Test
+    fun testLongDistribution() {
+        // Check that each bit is equally likely to be a 0 or 1.
+        
+        val count = 10000
+        val bitCounts = IntArray(64)
+        val rng = getRng()
+        for (i in 1..count) {
+            val value = rng.nextLong()
+            for (j in 0 until 64) {
+                if (value and (1L shl j) != 0L)
+                    bitCounts[j]++
+            }
+        }
+        val tol = 4*sqrt(count/2)
+        for (i in 0 until 64) {
+            assert(bitCounts[i] >= count/2 - tol)
+            assert(bitCounts[i] <= count/2 + tol)
+        }
+    }
+    
+    @Test
+    fun testBoundedIntDistribution() {
+        // Check that each value occurs equally often.
+
+        val count = 10000
+        val bins = 10
+        val binCounts = IntArray(bins)
+        val rng = getRng()
+        for (i in 1..count) {
+            val value = rng.nextInt(bins)
+            assert(value >= 0 && value < bins)
+            binCounts[value]++
+        }
+        val tol = 4*sqrt(count/bins)
+        for (i in 0 until bins) {
+            assert(binCounts[i] >= count/bins - tol)
+            assert(binCounts[i] <= count/bins + tol)
+        }
+    }
+    
+    @Test
+    fun testBoundedLongDistribution() {
+        // Check that each value occurs equally often.
+
+        val count = 10000
+        val bins = 10
+        val binCounts = IntArray(bins)
+        val rng = getRng()
+        for (i in 1..count) {
+            val value = rng.nextLong(bins.toLong())
+            assert(value >= 0 && value < bins)
+            binCounts[value.toInt()]++
+        }
+        val tol = 4*sqrt(count/bins)
+        for (i in 0 until bins) {
+            assert(binCounts[i] >= count/bins - tol)
+            assert(binCounts[i] <= count/bins + tol)
+        }
+    }
+
+    @Test
+    fun testFloatDistribution() {
+        // Check that values are uniformly distributed between 0 and 1.
+
+        val count = 10000
+        val bins = 10
+        val binCounts = IntArray(bins)
+        val rng = getRng()
+        for (i in 1..count) {
+            val value = rng.nextFloat()
+            assert(value >= 0.0f && value < 1.0f)
+            binCounts[(value*bins).toInt()]++
+        }
+        val tol = 4*sqrt(count/bins)
+        for (i in 0 until bins) {
+            assert(binCounts[i] >= count/bins - tol)
+            assert(binCounts[i] <= count/bins + tol)
```

Comment:
I can add more tests that check for specific values to make sure nothing has changed.  But for the statistical tests it's important to use a different seed every time (which they currently don't, but Kotlin 1.3 will fix that).  Otherwise you don't really know if the code works, only that the test passes for one particular seed.

Suppose you write a test that statistically should pass 99.9% of the time, but there's a bug that makes it fail 50% of the time.  If you use a different seed each time, you'll quickly discover the problem because the test will keep failing.  But if you use a fixed seed, there's a 50% chance you'll get unlucky and pick one for which the test passes.  And it will keep on passing forever and you'll never know the code is broken.

(Yes, I've made that mistake before and gotten bitten by it!)

Reply:
Sure, so the way I usually handle these cases is pick a set of static seeds (doesn't have to be one, could be a set of 100+ pre-picked numbers, but always the same ones) that I check on every commit (`koma-tests` runs in the CI every time a commit is pushed) and then have a truly random testing mode which can be manually enabled, perhaps with a `-PenableRandomTests`. When this gradle flag is set, the seeds are set completely randomly when it runs the random tests.

The reason to not allow truly randomized tests in the default test set that the CI runs (and that the user runs by default when they type `./gradlew test`) is to prevent someone's new commit to randomly fail because we happened to draw samples from a tail (i.e. there _is_ a valid output of a true rng that generates a thousand `1.0` values in a row, it's just very unlikely). I've chased down many a CI failure that seemed to be a bug in new code that turned out to just be unlikely seed selection in the test. The `enableRandomTests` mode can also run much longer (and have much lower tolerances) since it isn't the thing the user will be running by default.

We could also set the CI to run the truly randomized tests once a week, so that they aren't just forgotten and never used. That is, perform a periodic deep testing of the system.